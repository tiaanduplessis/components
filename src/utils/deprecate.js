/* eslint-disable no-console */
import {useRef, useCallback} from 'react'

const noop = () => {}
// eslint-disable-next-line import/no-mutable-exports
let deprecate = null

if (__DEV__) {
  deprecate = ({name, message, version}) => {
    Deprecations.deprecate({name, message, version})
  }
} else {
  deprecate = noop
}

export {deprecate}

// eslint-disable-next-line import/no-mutable-exports
let useDeprecation = null

if (__DEV__) {
  useDeprecation = ({name, message, version}) => {
    const ref = useRef(false)
    const logDeprecation = useCallback(() => {
      if (!ref.current) {
        ref.current = true
        deprecate({name, message, version})
      }
    }, [name, message, version])

    return logDeprecation
  }
} else {
  useDeprecation = () => {
    return noop
  }
}

export {useDeprecation}

export class Deprecations {
  static get() {
    if (!Deprecations.instance) {
      Deprecations.instance = new Deprecations()
    }

    return Deprecations.instance
  }

  constructor() {
    this.deprecations = []
  }

  static deprecate({name, message, version}) {
    const msg = `WARNING! ${name} is deprecated and will be removed in version ${version}. ${message}`
    console.warn(msg)

    this.get().deprecations.push({name, message, version})
  }

  static getDeprecations() {
    return this.get().deprecations
  }

  static clearDeprecations() {
    this.get().deprecations.length = 0
  }
}
