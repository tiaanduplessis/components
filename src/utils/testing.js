/* eslint-disable jsdoc/require-param, jsdoc/require-returns, jsdoc/require-description-complete-sentence, jsdoc/match-description */
import React from 'react'
import {promisify} from 'util'
import renderer from 'react-test-renderer'
import enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {ThemeProvider} from 'styled-components'
import {default as defaultTheme} from '../theme'

const readFile = promisify(require('fs').readFile)

export const COMPONENT_DISPLAY_NAME_REGEX = /^[A-Z][A-Za-z]+(\.[A-Z][A-Za-z]+)*$/

enzyme.configure({adapter: new Adapter()})

export function mount(component) {
  return enzyme.mount(component)
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * into its intermediate object representation with 'type',
 * 'props', and 'children' keys
 *
 * The returned object can be matched with expect().toEqual(), e.g.
 *
 * ```js
 * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
 * ```
 */
export function render(component) {
  return renderer.create(component).toJSON()
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * using react-test-renderer and return the root node
 * ```
 */
export function renderRoot(component) {
  return renderer.create(component).root
}

/**
 * Get the HTML class names rendered by the component instance
 * as an array.
 *
 * ```js
 * expect(renderClasses(<div className='a b' />))
 *   .toEqual(['a', 'b'])
 * ```
 */
export function renderClasses(component) {
  const {
    props: {className}
  } = render(component)
  return className ? className.trim().split(' ') : []
}

/**
 * Returns true if a node renders with a single class.
 */
export function rendersClass(node, klass) {
  return renderClasses(node).includes(klass)
}

export function renderWithTheme(node, theme = defaultTheme) {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>)
}

export function px(value) {
  return typeof value === 'number' ? `${value}px` : value
}

export function percent(value) {
  return typeof value === 'number' ? `${value}%` : value
}

export function renderStyles(node) {
  const {
    props: {className}
  } = render(node)
  return getComputedStyles(className)
}

export function getComputedStyles(className) {
  const div = document.createElement('div')
  div.className = className

  const computed = {}
  for (const sheet of document.styleSheets) {
    for (const rule of sheet.cssRules) {
      if (rule.type === 1) {
        readRule(rule, computed)
      } else if (rule.type === 4) {
        readMedia(rule)
      } else {
        // console.warn('rule.type =', rule.type)
      }
    }
  }

  return computed

  function readRule(rule, dest) {
    if (matchesSafe(div, rule.selectorText)) {
      const {style} = rule
      for (let i = 0; i < style.length; i++) {
        const prop = style[i]
        dest[prop] = style.getPropertyValue(prop)
      }
    } else {
      // console.warn('no match:', rule.selectorText)
    }
  }

  function readMedia(mediaRule) {
    const key = `@media ${mediaRule.media[0]}`
    // const dest = computed[key] || (computed[key] = {})
    const dest = {}
    for (const rule of mediaRule.cssRules) {
      readRule(rule, dest)
    }

    // Don't add media rule to computed styles
    // if no styles were actually applied
    if (Object.keys(dest).length > 0) {
      computed[key] = dest
    }
  }

  function matchesSafe(node, selector) {
    if (!selector) {
      return false
    }
    try {
      return node.matches(selector)
    } catch (error) {
      return false
    }
  }
}

/**
 * This provides a layer of compatibility between the render() function from
 * react-test-renderer and Enzyme's mount()
 */
export function getProps(node) {
  return typeof node.props === 'function' ? node.props() : node.props
}

export function getClassName(node) {
  return getProps(node).className
}

export function getClasses(node) {
  const className = getClassName(node)
  return className ? className.trim().split(/ +/) : []
}

export function loadCSS(path) {
  return readFile(require.resolve(path), 'utf8').then(css => {
    const style = document.createElement('style')
    style.setAttribute('data-path', path)
    style.textContent = css
    document.head.appendChild(style)
    return style
  })
}

export function unloadCSS(path) {
  const style = document.querySelector(`style[data-path="${path}"]`)
  if (style) {
    style.remove()
    return true
  }
}

// If a component requires certain props or other conditions in order
// to render without errors, you can pass a `toRender` function that
// returns an element ready to be rendered.
export function behavesAsComponent(Component, systemPropArray, toRender = null, options) {
  if (typeof toRender === 'object' && !options) {
    options = toRender
    toRender = null
  }
  options = options || {}

  const getElement = () => (toRender ? toRender() : <Component />)

  it('implements system props', () => {
    for (const systemProps of systemPropArray) {
      expect(Component).toImplementSystemProps(systemProps)
    }
  })

  if (!options.skipSx) {
    it('implements the sx prop', () => {
      expect(Component).toImplementSxProp()
    })

    it('implements sx prop behavior', () => {
      expect(getElement()).toImplementSxBehavior()
    })
  }

  if (!options.skipAs) {
    it('respects the as prop', () => {
      const As = React.forwardRef((_props, ref) => <div className="as-component" ref={ref} />)
      const elem = React.cloneElement(getElement(), {as: As})
      expect(render(elem)).toEqual(render(<As />))
    })
  }

  it('sets a valid displayName', () => {
    expect(Component.displayName).toMatch(COMPONENT_DISPLAY_NAME_REGEX)
  })

  it('sets the default theme', () => {
    expect(getElement()).toSetDefaultTheme()
  })

  it('renders consistently', () => {
    expect(render(getElement())).toMatchSnapshot()
  })
}

export function checkExports(path, exports) {
  it('has declared exports', () => {
    const mod = require(`../${path}`)
    expect(mod).toSetExports(exports)
  })
}
