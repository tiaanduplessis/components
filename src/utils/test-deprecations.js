import semver from 'semver'
import {Deprecations} from '../utils/deprecate'

const ourVersion = require('../../package.json').version

beforeEach(() => {
  Deprecations.clearDeprecations()
})

afterEach(() => {
  const deprecations = Deprecations.getDeprecations()

  for (const dep of deprecations) {
    if (semver.gte(ourVersion, dep.version)) {
      throw new Error(`Found a deprecation that should be removed in ${dep.version}`)
    }
  }
})
