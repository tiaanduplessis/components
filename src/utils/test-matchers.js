import React from 'react'
import 'jest-styled-components'
import {styleSheetSerializer} from 'jest-styled-components/serializer'
import theme from '../theme'
import {getClasses, mount, getComputedStyles, render} from './testing'

expect.addSnapshotSerializer(styleSheetSerializer)

const stringify = d => JSON.stringify(d, null, '  ')

/**
 * These are props that styled-system aliases for backwards compatibility.
 * For some reason, they don't show up in our toImplementSystemProps() matcher,
 * so we skip over them.
 */
const ALIAS_PROP_TYPES = ['w', 'align', 'justify', 'wrap']

expect.extend({
  toMatchKeys(obj, values) {
    return {
      pass: Object.keys(values).every(key => this.equals(obj[key], values[key])),
      message: () => `Expected ${stringify(obj)} to have matching keys: ${stringify(values)}`
    }
  },

  toHaveClass(node, klass) {
    const classes = getClasses(node)
    const pass = classes.includes(klass)
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klass)}`
    }
  },

  toHaveClasses(node, klasses, only = false) {
    const classes = getClasses(node)
    const pass = only ? this.equals(classes.sort(), klasses.sort()) : klasses.every(klass => classes.includes(klass))
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klasses)}`
    }
  },

  toImplementSystemProps(Component, propNames) {
    const propKeys = new Set(Object.keys(Component.propTypes))
    const expectedPropKeys = Object.keys(propNames.propTypes)
    const missing = expectedPropKeys.filter(key => !propKeys.has(key)).filter(key => !ALIAS_PROP_TYPES.includes(key))
    return {
      pass: missing.length === 0,
      message: () => `Missing prop${missing.length === 1 ? '' : 's'}: ${stringify(missing)}`
    }
  },

  toImplementSxProp(Component) {
    return {
      pass: !!Component.propTypes.sx,
      message: () => 'Missing sx propTypes'
    }
  },

  toImplementSxBehavior(element) {
    const mediaKey = '@media (max-width:123px)'
    const sxPropValue = {
      [mediaKey]: {
        color: 'red.5'
      }
    }

    const elem = React.cloneElement(element, {sx: sxPropValue})
    const rendered = render(elem)

    function checkStylesDeep(rendered) {
      const className = rendered.props ? rendered.props.className : ''
      const styles = getComputedStyles(className)
      if (styles[mediaKey] && styles[mediaKey].color) {
        return true
      } else if (rendered.children) {
        return rendered.children.some(child => checkStylesDeep(child))
      } else {
        return false
      }
    }

    return {
      pass: checkStylesDeep(rendered),
      message: () => 'sx prop values did not change styles of component nor of any sub-components'
    }
  },

  toSetDefaultTheme(Component) {
    let comp
    if (Component.type) {
      comp = Component
    } else {
      comp = <Component />
    }
    const wrapper = mount(comp)
    const pass = this.equals(wrapper.prop('theme'), theme)
    return {
      pass,
      message: () => 'default theme is not set'
    }
  },

  toSetExports(mod, expectedExports) {
    if (!Object.keys(expectedExports).includes('default')) {
      return {
        pass: false,
        message: () => "You must specify the module's default export"
      }
    }

    const seen = new Set()
    for (const exp of Object.keys(expectedExports)) {
      seen.add(exp)
      if (mod[exp] !== expectedExports[exp]) {
        if (!mod[exp] && !expectedExports[exp]) {
          continue
        }

        return {
          pass: false,
          message: () => `Module exported a different value from key '${exp}' than expected`
        }
      }
    }

    for (const exp of Object.keys(mod)) {
      if (seen.has(exp)) {
        continue
      }

      if (mod[exp] !== expectedExports[exp]) {
        return {
          pass: false,
          message: () => `Module exported an unexpected value from key '${exp}'`
        }
      }
    }

    return {
      pass: true,
      message: () => ''
    }
  }
})
