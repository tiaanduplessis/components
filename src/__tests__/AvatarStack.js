import React from 'react'
import {AvatarStack} from '..'
import {render, behavesAsComponent, checkExports} from '../utils/testing'
import {COMMON} from '../constants'
import {render as HTMLRender, cleanup} from '@testing-library/react'
import {axe, toHaveNoViolations} from 'jest-axe'
import 'babel-polyfill'
expect.extend(toHaveNoViolations)

const avatarComp = (
  <AvatarStack>
    <img src="https://avatars.githubusercontent.com/primer" alt="" />
    <img src="https://avatars.githubusercontent.com/github" alt="" />
    <img src="https://avatars.githubusercontent.com/primer" alt="" />
    <img src="https://avatars.githubusercontent.com/github" alt="" />
  </AvatarStack>
)

const rightAvatarComp = (
  <AvatarStack alignRight>
    <img src="https://avatars.githubusercontent.com/primer" alt="" />
    <img src="https://avatars.githubusercontent.com/github" alt="" />
    <img src="https://avatars.githubusercontent.com/primer" alt="" />
    <img src="https://avatars.githubusercontent.com/github" alt="" />
  </AvatarStack>
)

describe('Avatar', () => {
  behavesAsComponent(AvatarStack, [COMMON], () => avatarComp)

  checkExports('AvatarStack', {
    default: AvatarStack
  })

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(avatarComp)
    const results = await axe(container)
    expect(results).toHaveNoViolations()
    cleanup()
  })

  it('respects alignRight props', () => {
    expect(render(rightAvatarComp)).toMatchSnapshot()
  })
})
