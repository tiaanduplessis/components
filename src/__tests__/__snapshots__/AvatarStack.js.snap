// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders consistently 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 38px;
  position: absolute;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 20px;
  min-width: 38px;
}

.c0 .pc-AvatarItem {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 20px;
  width: 20px;
  box-shadow: 0 0 0 1px #fff;
  margin-left: -11px;
  position: relative;
  overflow: hidden;
  -webkit-transition: margin 0.2s ease-in-out,opacity 0.2s ease-in-out,visibility 0.2s ease-in-out,box-shadow 0.1s ease-in-out;
  transition: margin 0.2s ease-in-out,opacity 0.2s ease-in-out,visibility 0.2s ease-in-out,box-shadow 0.1s ease-in-out;
}

.c0 .pc-AvatarItem:first-child {
  margin-left: 0;
}

.c0 .pc-AvatarItem:nth-child(n + 4) {
  display: none;
}

.c0.pc-AvatarStack--two {
  min-width: 30px;
}

.c0.pc-AvatarStack--two .pc-AvatarItem:nth-child(n + 3) {
  display: none;
}

.c0.pc-AvatarStack--three-plus {
  min-width: 38px;
}

.c0.pc-AvatarStack--three-plus .pc-AvatarItem:nth-child(3) {
  opacity: 55%;
  margin-left: -17px;
}

.c0.pc-AvatarStack--three-plus .pc-AvatarItem:nth-child(4) {
  opacity: 40%;
  margin-left: -17px;
}

.c0.pc-AvatarStack--three-plus .pc-AvatarItem:nth-child(5) {
  opacity: 25%;
  margin-left: -17px;
}

.c0.pc-AvatarStack--three-plus .pc-AvatarItem:nth-child(n + 4) {
  display: block;
}

.c0.pc-AvatarStack--three-plus .pc-AvatarItem:nth-child(n + 6) {
  opacity: 0;
  visibility: hidden;
}

.c0.pc-AvatarStack--right {
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c0.pc-AvatarStack--right .pc-AvatarItem {
  margin-left: 0 !important;
  margin-right: -11px;
}

.c0.pc-AvatarStack--right .pc-AvatarItem:first-child {
  margin-right: 0;
}

.c0.pc-AvatarStack--right .pc-AvatarStackBody {
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c0.pc-AvatarStack--right .pc-AvatarStackBody:hover .pc-AvatarItem {
  margin-right: 4px!important;
  margin-left: 0 !important;
}

.c0.pc-AvatarStack--right .pc-AvatarStackBody:hover .pc-AvatarItem:first-child {
  margin-right: 0 !important;
}

.c0.pc-AvatarStack--three-plus.pc-AvatarStack--right .pc-AvatarItem:nth-child(3) {
  margin-right: -17px;
}

.c0.pc-AvatarStack--three-plus.pc-AvatarStack--right .pc-AvatarItem:nth-child(4) {
  margin-right: -17px;
}

.c0.pc-AvatarStack--three-plus.pc-AvatarStack--right .pc-AvatarItem:nth-child(5) {
  margin-right: -17px;
}

.c0 .pc-AvatarStackBody:hover {
  width: auto;
}

.c0 .pc-AvatarStackBody:hover .pc-AvatarItem {
  margin-left: 4px;
  opacity: 100%;
  visibility: visible;
  box-shadow: 0 0 0 4px #fff;
}

.c0 .pc-AvatarStackBody:hover .pc-AvatarItem:first-child {
  margin-left: 0;
}

<span
  className="c0 pc-AvatarStack--three-plus"
>
  <div
    className="c1 pc-AvatarStackBody"
    display="flex"
    width="38px"
  >
    <img
      alt=""
      className="pc-AvatarItem"
      src="https://avatars.githubusercontent.com/primer"
      sx={
        Object {
          "zIndex": 10,
        }
      }
    />
    <img
      alt=""
      className="pc-AvatarItem"
      src="https://avatars.githubusercontent.com/github"
      sx={
        Object {
          "zIndex": 9,
        }
      }
    />
    <img
      alt=""
      className="pc-AvatarItem"
      src="https://avatars.githubusercontent.com/primer"
      sx={
        Object {
          "zIndex": 8,
        }
      }
    />
    <img
      alt=""
      className="pc-AvatarItem"
      src="https://avatars.githubusercontent.com/github"
      sx={
        Object {
          "zIndex": 7,
        }
      }
    />
  </div>
</span>
`;

exports[`Avatar respects alignRight props 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 38px;
  position: absolute;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 20px;
  min-width: 38px;
}

.c0 .pc-AvatarItem {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 20px;
  width: 20px;
  box-shadow: 0 0 0 1px #fff;
  margin-left: -11px;
  position: relative;
  overflow: hidden;
  -webkit-transition: margin 0.2s ease-in-out,opacity 0.2s ease-in-out,visibility 0.2s ease-in-out,box-shadow 0.1s ease-in-out;
  transition: margin 0.2s ease-in-out,opacity 0.2s ease-in-out,visibility 0.2s ease-in-out,box-shadow 0.1s ease-in-out;
}

.c0 .pc-AvatarItem:first-child {
  margin-left: 0;
}

.c0 .pc-AvatarItem:nth-child(n + 4) {
  display: none;
}

.c0.pc-AvatarStack--two {
  min-width: 30px;
}

.c0.pc-AvatarStack--two .pc-AvatarItem:nth-child(n + 3) {
  display: none;
}

.c0.pc-AvatarStack--three-plus {
  min-width: 38px;
}

.c0.pc-AvatarStack--three-plus .pc-AvatarItem:nth-child(3) {
  opacity: 55%;
  margin-left: -17px;
}

.c0.pc-AvatarStack--three-plus .pc-AvatarItem:nth-child(4) {
  opacity: 40%;
  margin-left: -17px;
}

.c0.pc-AvatarStack--three-plus .pc-AvatarItem:nth-child(5) {
  opacity: 25%;
  margin-left: -17px;
}

.c0.pc-AvatarStack--three-plus .pc-AvatarItem:nth-child(n + 4) {
  display: block;
}

.c0.pc-AvatarStack--three-plus .pc-AvatarItem:nth-child(n + 6) {
  opacity: 0;
  visibility: hidden;
}

.c0.pc-AvatarStack--right {
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c0.pc-AvatarStack--right .pc-AvatarItem {
  margin-left: 0 !important;
  margin-right: -11px;
}

.c0.pc-AvatarStack--right .pc-AvatarItem:first-child {
  margin-right: 0;
}

.c0.pc-AvatarStack--right .pc-AvatarStackBody {
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c0.pc-AvatarStack--right .pc-AvatarStackBody:hover .pc-AvatarItem {
  margin-right: 4px!important;
  margin-left: 0 !important;
}

.c0.pc-AvatarStack--right .pc-AvatarStackBody:hover .pc-AvatarItem:first-child {
  margin-right: 0 !important;
}

.c0.pc-AvatarStack--three-plus.pc-AvatarStack--right .pc-AvatarItem:nth-child(3) {
  margin-right: -17px;
}

.c0.pc-AvatarStack--three-plus.pc-AvatarStack--right .pc-AvatarItem:nth-child(4) {
  margin-right: -17px;
}

.c0.pc-AvatarStack--three-plus.pc-AvatarStack--right .pc-AvatarItem:nth-child(5) {
  margin-right: -17px;
}

.c0 .pc-AvatarStackBody:hover {
  width: auto;
}

.c0 .pc-AvatarStackBody:hover .pc-AvatarItem {
  margin-left: 4px;
  opacity: 100%;
  visibility: visible;
  box-shadow: 0 0 0 4px #fff;
}

.c0 .pc-AvatarStackBody:hover .pc-AvatarItem:first-child {
  margin-left: 0;
}

<span
  className="c0 pc-AvatarStack--three-plus pc-AvatarStack--right"
>
  <div
    className="c1 pc-AvatarStackBody"
    display="flex"
    width="38px"
  >
    <img
      alt=""
      className="pc-AvatarItem"
      src="https://avatars.githubusercontent.com/primer"
      sx={
        Object {
          "zIndex": 10,
        }
      }
    />
    <img
      alt=""
      className="pc-AvatarItem"
      src="https://avatars.githubusercontent.com/github"
      sx={
        Object {
          "zIndex": 9,
        }
      }
    />
    <img
      alt=""
      className="pc-AvatarItem"
      src="https://avatars.githubusercontent.com/primer"
      sx={
        Object {
          "zIndex": 8,
        }
      }
    />
    <img
      alt=""
      className="pc-AvatarItem"
      src="https://avatars.githubusercontent.com/github"
      sx={
        Object {
          "zIndex": 7,
        }
      }
    />
  </div>
</span>
`;
